tests-integration/src/tests/instance_kafka_wal_test.rs

```diff
@@ -74,24 +82,196 @@ async fn test_create_database_and_insert_query(instance: Option<Box<dyn Rebuilda
+    assert_matches!(output, Output::AffectedRows(1));
+
+    let tables = create_tables("test_replay", &instance, 3).await;
+    insert_data(&tables, &instance, 5).await;
+    ensure_data_exists(&tables, &instance).await;
+
+    // Rebuilds to emulate restart which then triggers a replay.
+    let instance = rebuildable_instance.rebuild().await;
+    ensure_data_exists(&tables, &instance).await;
 }
 
-async fn execute_sql_with(
+#[apply(both_instances_cases_with_kafka_wal)]
+async fn test_flush_then_replay(rebuildable_instance: Option<Box<dyn RebuildableMockInstance>>) {
+    let Some(mut rebuildable_instance) = rebuildable_instance else {
+        return;
+    };
+    let instance = rebuildable_instance.frontend();
+
+    let output = execute_sql(&instance, "create database test").await;
+    assert_matches!(output, Output::AffectedRows(1));
+
+    let tables = create_tables("test_flush_then_replay", &instance, 3).await;
+    insert_data(&tables, &instance, 5).await;
+    ensure_data_exists(&tables, &instance).await;
+
+    // Renames tables to force flushing each table.
+    let tables = futures::future::join_all(tables.into_iter().map(|table| {
+        let instance = instance.clone();
+        async move {
+            // Repeats the last char to construct a new table name.
+            let new_table_name = format!("{}{}", table.name, table.name.chars().last().unwrap());
+            assert_matches!(
+                do_alter(&instance, &table.name, &new_table_name).await,
+                Output::AffectedRows(0)
+            );
+            Arc::new(Table {
+                name: new_table_name,
+                logical_timer: AtomicU64::new(table.logical_timer.load(Ordering::Relaxed)),
+                inserted: Mutex::new(table.inserted.lock().await.clone()),
+            })
+        }
+    }))
+    .await;
+    ensure_data_exists(&tables, &instance).await;
+
+    // Rebuilds to emulate restart which then triggers a replay.
+    let instance = rebuildable_instance.rebuild().await;
+    ensure_data_exists(&tables, &instance).await;
+}
+
+async fn create_tables(
+    test_name: &str,
     instance: &Arc<Instance>,
-    sql: &str,
-    query_ctx: QueryContextRef,
-) -> Output {
-    try_execute_sql_with(instance, sql, query_ctx)
+    num_tables: usize,
+) -> Vec<Arc<Table>> {
+    futures::future::join_all((0..num_tables).map(|i| {
+        let instance = instance.clone();
+        async move {
+            let table_name = format!("{}_{}", test_name, i);
+            assert_matches!(
+                do_create(&instance, &table_name).await,
+                Output::AffectedRows(0)
+            );
+            Arc::new(Table {
+                name: table_name,
+                logical_timer: AtomicU64::new(1685508715000),
+                inserted: Mutex::new(Vec::new()),
+            })
+        }
+    }))
+    .await
+}
+
+async fn insert_data(tables: &[Arc<Table>], instance: &Arc<Instance>, num_writers: usize) {
```

Comment:
You can change it to mutable and without a lock
```suggestion
async fn insert_data(tables: &mut [Table], instance: &Arc<Instance>, num_writers: usize)
```